From 32a7b1888429d74abeebaa4b3936f2959afde4d8 Mon Sep 17 00:00:00 2001
From: songliyong <songliyong@contributor>


Date: Mon, 22 Dec 2025 17:06:11 +0800
Subject: [PATCH] BUGFIX fix temp table offset pushdown issue

Issue:
----------
DestroyTuple crashed in offset pushdown.
We can not destroy tuple for temp table.

Solution:
----------
Skip destroy tuple if it's a temp table.
---
 .../dstore_main/r/offset_pushdown-dstore.result     | 13 +++++++++++++
 .../suite/dstore_main/t/offset_pushdown-dstore.test |  9 +++++++++
 storage/dstore/handler/ha_cde.cc                    |  4 +++-
 3 files changed, 25 insertions(+), 1 deletion(-)

diff --git a/mysql-test/suite/dstore_main/r/offset_pushdown-dstore.result b/mysql-test/suite/dstore_main/r/offset_pushdown-dstore.result
index 15ad83621ae..6af4be0624f 100644
--- a/mysql-test/suite/dstore_main/r/offset_pushdown-dstore.result
+++ b/mysql-test/suite/dstore_main/r/offset_pushdown-dstore.result
@@ -1232,3 +1232,16 @@ id	name	department_id	salary
 3	Bob	10	7000.00
 SET optimizer_switch = @orig_optimizer_switch;
 DROP table employees;
+CREATE TABLE non_tmp(a int, b int);
+INSERT INTO non_tmp values(1,1),(2,1),(3,1),(4,1),(5,1);
+SELECT * FROM non_tmp limit 2 offset 1 FOR UPDATE;
+a	b
+2	1
+3	1
+CREATE TEMPORARY TABLE tmp_tbl(a int, b int);
+INSERT INTO tmp_tbl select * from non_tmp;
+SELECT * FROM tmp_tbl limit 2 offset 1 FOR UPDATE;
+a	b
+2	1
+3	1
+DROP table non_tmp, tmp_tbl;
diff --git a/mysql-test/suite/dstore_main/t/offset_pushdown-dstore.test b/mysql-test/suite/dstore_main/t/offset_pushdown-dstore.test
index 642c9629b39..13ac85ebe60 100644
--- a/mysql-test/suite/dstore_main/t/offset_pushdown-dstore.test
+++ b/mysql-test/suite/dstore_main/t/offset_pushdown-dstore.test
@@ -649,3 +649,12 @@ EXPLAIN  SELECT * FROM employees  WHERE department_id <30 limit 2 offset 1;
 SELECT * FROM employees  WHERE department_id < 30 limit 2 offset 1;
 SET optimizer_switch = @orig_optimizer_switch;
 DROP table employees;
+
+CREATE TABLE non_tmp(a int, b int);
+INSERT INTO non_tmp values(1,1),(2,1),(3,1),(4,1),(5,1);
+SELECT * FROM non_tmp limit 2 offset 1 FOR UPDATE;
+
+CREATE TEMPORARY TABLE tmp_tbl(a int, b int);
+INSERT INTO tmp_tbl select * from non_tmp;
+SELECT * FROM tmp_tbl limit 2 offset 1 FOR UPDATE;
+DROP table non_tmp, tmp_tbl;
\ No newline at end of file
diff --git a/storage/dstore/handler/ha_cde.cc b/storage/dstore/handler/ha_cde.cc
index 2c478c59238..7f550e33442 100644
--- a/storage/dstore/handler/ha_cde.cc
+++ b/storage/dstore/handler/ha_cde.cc
@@ -3586,7 +3586,9 @@ int ha_cde::rnd_next(
     if (skip_row_for_offset_pushdown()) {
       increment_scanned_offset();
       if (m_dstore->select_lock_type != LOCK_NONE && tuple) {
-        TupleInterface::DestroyTuple(tuple, CdeFreeMemForDtuple);
+        if (!m_dstore->table_handler->is_temporary()) {
+          TupleInterface::DestroyTuple(tuple, CdeFreeMemForDtuple);
+        }
         tuple = nullptr;
       }
       continue;
-- 
2.23.0

